deepmd/dpmodel/descriptor/dpa1.py

```diff
@@ -0,0 +1,911 @@
+            "hidden_dim": self.hidden_dim,
+            "dotr": self.dotr,
+            "do_mask": self.do_mask,
+            "scaling_factor": self.scaling_factor,
+            "normalize": self.normalize,
+            "temperature": self.temperature,
+            "precision": self.precision,
+            "attention_layer": self.attention_layer.serialize(),
+            "attn_layer_norm": self.attn_layer_norm.serialize(),
+        }
+
+    @classmethod
+    def deserialize(cls, data) -> "NeighborGatedAttentionLayer":
+        """Deserialize the networks from a dict.
+
+        Parameters
+        ----------
+        data : dict
+            The dict to deserialize from.
+        """
+        data = data.copy()
+        attention_layer = data.pop("attention_layer")
+        attn_layer_norm = data.pop("attn_layer_norm")
+        obj = cls(**data)
+        obj.attention_layer = GatedAttentionLayer.deserialize(attention_layer)
+        obj.attn_layer_norm = LayerNorm.deserialize(attn_layer_norm)
+        return obj
+
+
+class GatedAttentionLayer(NativeOP):
+    def __init__(
+        self,
+        nnei: int,
+        embed_dim: int,
+        hidden_dim: int,
+        dotr: bool = False,
+        do_mask: bool = False,
+        scaling_factor: float = 1.0,
+        normalize: bool = True,
+        temperature: Optional[float] = None,
+        bias: bool = True,
+        smooth: bool = True,
+        precision: str = DEFAULT_PRECISION,
+    ):
+        """Construct a neighbor-wise attention net."""
+        super().__init__()
+        self.nnei = nnei
+        self.embed_dim = embed_dim
+        self.hidden_dim = hidden_dim
+        self.dotr = dotr
+        self.do_mask = do_mask
+        self.bias = bias
+        self.smooth = smooth
+        self.scaling_factor = scaling_factor
+        self.temperature = temperature
+        self.precision = precision
+        if temperature is None:
+            self.scaling = (self.hidden_dim * scaling_factor) ** -0.5
+        else:
+            self.scaling = temperature
+        self.normalize = normalize
+        self.in_proj = NativeLayer(
+            embed_dim,
+            hidden_dim * 3,
+            bias=bias,
+            use_timestep=False,
+            precision=precision,
+        )
+        self.out_proj = NativeLayer(
+            hidden_dim,
+            embed_dim,
+            bias=bias,
+            use_timestep=False,
+            precision=precision,
+        )
+
+    def call(self, query, nei_mask, input_r=None, sw=None, attnw_shift=20.0):
+        # Linear projection
+        q, k, v = np.split(self.in_proj(query), 3, axis=-1)
+        # Reshape and normalize
+        q = q.reshape(-1, self.nnei, self.hidden_dim)
+        k = k.reshape(-1, self.nnei, self.hidden_dim)
+        v = v.reshape(-1, self.nnei, self.hidden_dim)
+        if self.normalize:
+            q = np_normalize(q, axis=-1)
+            k = np_normalize(k, axis=-1)
+            v = np_normalize(v, axis=-1)
```

Comment:
really need to normalize q, k and v?

Reply:
When developing dpa1, it was found that without normalization here, the attention weights are often dominated by `q` and `k` values from the largest `g_ij` (typically the nearest neighbor). This can result in the attention mechanism failing to capture relevant dependencies among other neighbors. 

I also tried adding normalization in various places, and found that applying it here helps stabilize the attention weights and training process.

This phenomenon is likely due to the fact that our `g_ij` uses distance scale information as input, which differs from the typical attention mechanisms in NLP. Of course, there are now more changes in other parts of the system, so if have time, we may explore other potential solutions as well.

Reply:
Interesting, we may reconsider the attentions in dpa2, which did not use this normalization.